kernel/typing.ml

```diff
@@ -152,6 +152,56 @@ let unshift_reduce sg q t =
     ( try Some (Subst.unshift q (R.snf sg t))
       with Subst.UnshiftExn -> None )
 
+exception VarSurelyOccurs
+
+(** Under [d] lambdas, checks whether term [te] *must* contain an occurence
+    of any variable that satisfies the given predicate [p],
+    *even when substituted or reduced*.
+    This check make no assumption on the rewrite system or possible substitution
+    - any definable symbol are "safe" as they may reduce to a term where no variable occur
+    - any applied meta variable (DB index > [d]) are "safe" as they may be
+      substituted and the reduce to a term where no variable occur
+    Raises VarSurelyOccurs if the term [te] *surely* contains an occurence of one
+    of the [vars].
+ *)
+let sure_occur_check sg (d:int) (p:int -> bool) : term -> unit =
+  let rec aux k = function (* k counts the number of local lambda abstractions *)
+    | Kind | Type _ -> ()
+    | Const _ -> ()
+    | Pi  (_,_,     a,b) -> ( aux k a; aux (k+1) b )
+    | Lam (_,_,None  ,b) -> (          aux (k+1) b )
+    | Lam (_,_,Some a,b) -> ( aux k a; aux (k+1) b )
+    | DB (_,_,n) -> if n >= k && p (n-k) then raise VarSurelyOccurs
+    | App (f,a,args) ->
+      begin
+        match f with
+        | DB (_,_,n) when n >= k + d -> (* a matching variable *)
+          if p (n-k) then raise VarSurelyOccurs
+        | DB (_,_,n) when n < k + d -> (* a locally bound variable *)
+          if n >= k && p (n-k)
+          then raise VarSurelyOccurs
+          else ( aux k a; List.iter (aux k) args )
+        | Const (l,cst) when Signature.is_static sg l cst ->
+          ( aux k a; List.iter (aux k) args )
+        | _ -> ()
+        (* Default case encompasses:
+           - Meta variables: DB(_,_,n) with n >= k + d
+           - Definable symbols
+           - Lambdas (FIXME: when can this happen ?)
+           - Illegal applications  *)
+      end
+  in aux 0
+
+let gather_free_vars (d:int) (vars:bool array) : term -> unit =
```

Comment:
Efficiency is not a real issue here, but you could make this function recursive terminal.

Reply:
Done. It's a bit less readable but ok.